packages/network/src/connection/NodeWebRtcConnection.ts

```diff
@@ -0,0 +1,318 @@
+
+        this.doClose()
+    }
+
+    getBufferedAmount(): number {
+        try {
+            return this.dataChannel!.bufferedAmount().valueOf()
+        } catch (err) {
+            return 0
+        }
+    }
+
+    getMaxMessageSize(): number {
+        try {
+            return this.dataChannel!.maxMessageSize().valueOf()
+        } catch (err) {
+            return 1024 * 1024
+        }
+    }
+ 
+    isOpen(): boolean {
+        try {
+            return this.dataChannel!.isOpen()
+        } catch (err) {
+            return false
+        }
+    }
+
+    private onStateChange(state: string): void {
+        this.logger.trace('conn.onStateChange: %s -> %s', this.lastState, state)
+
+        this.lastState = state
+
+        if (state === 'disconnected' || state === 'closed') {
+            this.close()
+        } else if (state === 'failed') {
+            this.close(new Error('connection failed'))
+        } else if (state === 'connecting') {
+            // reset timeout on connecting
+            clearTimeout(this.connectionTimeoutRef!)
+            this.connectionTimeoutRef = setTimeout(() => {
+                if (this.isFinished) { return }
+                this.logger.warn(`connection timed out after ${this.newConnectionTimeout}ms`)
+                this.close(new Error(`timed out after ${this.newConnectionTimeout}ms`))
+            }, this.newConnectionTimeout)
+        }
+    }
+
+    private onGatheringStateChange(state: string): void {
+        this.logger.trace('conn.onGatheringStateChange: %s -> %s', this.lastGatheringState, state)
+        this.lastGatheringState = state
+    }
+
+    private onDataChannel(dataChannel: DataChannel): void {
+        this.setupDataChannel(dataChannel)
+        this.logger.trace('connection.onDataChannel')
+        this.openDataChannel(dataChannel)
+    }
+
+    private onLocalDescription(description: string, type: DescriptionType): void {
+        this.emit('localDescription', type, description)
+    }
+
+    private onLocalCandidate(candidate: string, mid: string): void {
+        this.emit('localCandidate', candidate, mid)
+    }
+
+    private setupDataChannel(dataChannel: DataChannel): void {
+        this.paused = false
+        this.dataChannelEmitter = DataChannelEmitter(dataChannel)
+        dataChannel.setBufferedAmountLowThreshold(this.bufferThresholdLow)
+        this.dataChannelEmitter.on('open', () => {
+            this.logger.trace('dc.onOpen')
+            this.openDataChannel(dataChannel)
+        })
+
+        this.dataChannelEmitter.on('closed', () => {
+            this.logger.trace('dc.onClosed')
+            this.close()
+        })
+
+        this.dataChannelEmitter.on('error', (err) => {
+            this.logger.warn('dc.onError: %s', err)
+        })
+
+        this.dataChannelEmitter.on('bufferedAmountLow', () => {
+            if (!this.paused) { return }
+            this.paused = false
+            this.setFlushRef()
+            this.emit('bufferLow')
```

Comment:
Again consider whether buffer handling logic should be shared or subclass-specific.